distribution/lib/Standard/Test/0.0.0-dev/src/Test.enso

```diff
@@ -167,3 +171,39 @@ type Test
         result = behavior
         State.put Clue prev_clue
         result
+
+    ## A helper method that retries the action a few times if it panics.
+       It allows to make flaky tests more robust.
+
+       It waits for a short period of time between retries (in case the failures
+       are related e.g. to network conditions or propagation delays).
+
+       This function should be placed inside of the `specify` block. It can be
+       used to retry the whole test, or it can be applied to a specific block
+       inside of the test to only re-run that specific block (in such case, any
+       side-effects must be considered carefully).
+
+       The method returns the first successful value returned by the action,
+       or fails with the last error thrown by the action after exhausting retry
+       attempts.
+    with_retries : Any -> Any
+    with_retries ~action =
+        loc = Meta.get_source_location 1
+
+        # Delays are in seconds
+        sleep_time = 0.1
+        total_sleep_delay = if Environment.get "CI" . is_nothing . not then 10 else 2
+
+        max_iterations = total_sleep_delay / sleep_time
+        go i =
+            Panic.catch Any action caught_panic->
+                # If the iterations are exhausted, we rethrow the panic.
+                if i > max_iterations then Panic.throw caught_panic else
+                    if i % 10 == 0 then
+                        IO.println "Still failing after "+i.to_text+" retries. ("+loc.to_display_text+")"
+                    Thread.sleep (1000*sleep_time . floor)
```

Comment:
The total time spent might be significantly longer if the `action` itself takes non-negligible time; it might be better to check the current time against `(start_time + total_sleep_delay)` rather than using a counter.

Reply:
Fair point, but I feel like the current behaviour is what we want.

If the action takes 3s to complete due to bad network conditions _and it fails on a timeout_, then with a retry delay of 2s - it will not retry at all... But the whole point of this is to do _some_ retries. I think it's better to do the same number of retries regardless of how long the underlying action is taking.

The `total_sleep_delay` is just used to approximate the total wait time. But I guess I can rephrase this to just be `max_retries` counter and remove the `total_sleep_delay` altogether.

Reply:
What about a 2-linear backoff? E.g., first retry waits for 2 seconds, another for 4 seconds, another for 8 seconds, 16 secs, etc.... The way you coded it, it will wait for 100 seconds on the CI after every retry, right?

Reply:
> What about a 2-linear backoff? E.g., first retry waits for 2 seconds, another for 4 seconds, another for 8 seconds, 16 secs, etc.... The way you coded it, it will wait for 100 seconds on the CI after every retry, right?

It will wait for 100 **milli**seconds between every retry, not 100 seconds 😅 

I feel like this is unnecessarily complicating stuff. I want the test to finish as soon as possible, so increasing the wait time does not seem to make that better. The strategy we have here was already successfully used for running cloud tests with propagation delays. I don't think there's value in complicating this strategy _until_ we have a reason to do so. For now, I don't see any reasons - it works good enough and is simple.